Improve LB interface testing

jngrad opened the pull request:
Description of changes:
- rename `print_vtk_*` functions to `write_vtk_*` functions in LB and EK to better reflect what they actually do
- simplify VTK bounding box code and make it coherent with the numpy slice syntax (i.e. `bb2` now needs zero-based indices)
- add assertions to catch an out of bounds bug in VTK (incorrect bounding box values caused the python interpreter to freeze)
- fix broken LB getters in the core and python interface
- fix incorrect LB density getter/setter check in the testsuite
- add extra checks for LB parameters
- add testcase for VTK output

jngrad committed:
tests: Check LB interface setter/getters and exceptions

jngrad committed:
core: Add guard for VTK output

Accessing nodes out of bounds causes ESPResSo to freeze.

jngrad committed:
python: Add docs, throw on invalid VTK bounding box

jngrad committed:
tests: Add testcase for VTK functions

Backport testcase from the 4.3 branch.

jngrad committed:
tests: Add testcase for non-VTK write functions

Also thermalize the fluid and document the non-VTK write functions.

KaiSzuttor commented on src/core/grid_based_algorithms/lb_interface.cpp:
```diff
@@ -563,28 +564,36 @@ void lb_lbfluid_print_vtk_velocity(const std::string &filename,
-    bb_high.push_back(std::max(*val1, *val2));
+    auto const lower = std::min(*val1, *val2);
+    auto const upper = std::max(*val1, *val2);
+    if (lower < 0) {
+      throw std::runtime_error("Cannot access index " + std::to_string(lower));
+    }
+    if (upper >= bb_high[it]) {
+      throw std::runtime_error("Cannot access index " + std::to_string(upper));
+    }
```
```suggestion
    if (lower < 0 or upper >= bb_high[it]) {
      throw std::runtime_error("Cannot access index " + std::to_string(lower));
    }
```

jngrad replied:
wouldn't the error message be confusing if `upper` is the one that causes the problem?

KaiSzuttor replied:
yes of course, overlooked that, but i think it would be best to merge those errors and just print the whole information: lower, upper, and the limits

KaiSzuttor commented on src/core/grid_based_algorithms/lb_interface.cpp:
```diff
@@ -563,28 +564,36 @@ void lb_lbfluid_print_vtk_velocity(const std::string &filename,
 
-  std::vector<int> bb_low;
-  std::vector<int> bb_high;
+  Utils::Vector3i bb_low{0, 0, 0};
+  Utils::Vector3i bb_high{0, 0, 0};
+  if (lattice_switch == ActiveLB::GPU) {
+#ifdef CUDA
+    bb_high = {static_cast<int>(lbpar_gpu.dim_x) - 1,
+               static_cast<int>(lbpar_gpu.dim_y) - 1,
+               static_cast<int>(lbpar_gpu.dim_z) - 1};
+#endif //  CUDA
+  } else {
+    bb_high = {lblattice.global_grid[0] - 1, lblattice.global_grid[1] - 1,
+               lblattice.global_grid[2] - 1};
+  }
```
```suggestion
  auto const bb_low = Utils::Vector3i{};
  auto const bb_high = lb_lbfluid_get_shape() - Utils::Vector3i{1,1,1};
```

KaiSzuttor replied:
here, we can probably drop the - 1 and just check for smaller

jngrad replied:
good idea!

jngrad replied:
done

KaiSzuttor commented on src/python/espressomd/lb.pyx:
```diff
@@ -205,23 +205,69 @@ cdef class HydrodynamicInteraction(Actor):
         lb_lbfluid_print_vtk_boundary(utils.to_char_pointer(path))
 
     def print_velocity(self, path):
+        """Write the LB fluid velocity to a data file that can be loaded by
```
we should consider renaming those functions because we actually write to a file and not just print to screen

jngrad replied:
maybe `write_velocity`? Also, do you know why the CPU LB implementation of `print_boundary` casts numbers to 1, while the GPU LB writes some integers?

KaiSzuttor replied:
i don't only mean this function but also the other ```print_*``` functions in this file. ```write_*``` would be a more accurate description

jngrad replied:
ok, but there are other places in the code with similarly confusing names, e.g. the electrokinetics `print_vtk_*` functions. Is it worth introducing this breaking change in 4.2, considering these LB and EK functions have already been removed in 4.3?

KaiSzuttor replied:
don't know, i think we should clean up what we stumble over regardless of the lifetime of the change

jngrad replied:
done

KaiSzuttor commented on testsuite/python/lb_vtk.py:
```diff
@@ -0,0 +1,225 @@
+try:
+    import vtk
+    from vtk.util import numpy_support as VN
+    skipIfMissingPythonPackage = ut.case._id
```
what's that?

jngrad replied:
`vtk` is not a hard dependency

KaiSzuttor replied:
but what is ```_id```

KaiSzuttor replied:
not sure if it's a good idea to use a private variable

jngrad replied:
that's just a lambda that returns the input function, it's called `_id` because `id` is a python built-in function

KaiSzuttor replied:
yes but if it was part of the official interface they'd looked for a proper name, right?

jngrad replied:
we now use our own `_id` lambda in 897c5bb118

KaiSzuttor reviewed (dismissed):
thanks for those tests! just minor improvements should be done before merging

jngrad committed:
tests: Use our _id() instead of unittest.case._id()

jngrad committed:
core: Simplify VTK velocity function

jngrad committed:
python: Reame print_vtk_* methods to write_vtk_*

This change better reflects what is actually carried out by these
functions. Affects LB and EK modules.

KaiSzuttor commented on testsuite/python/h5md.py:
```diff
@@ -30,7 +30,7 @@
 try:
     import h5py  # h5py has to be imported *after* espressomd (MPI)
-    skipIfMissingPythonPackage = ut.case._id
+    skipIfMissingPythonPackage = utx._id
```
okay, i still don't get why this should be a private function

jngrad replied:
I've renamed it to `no_skip` in 08a674fae

KaiSzuttor commented on testsuite/python/mdanalysis.py:
```diff
@@ -24,10 +24,11 @@
     import MDAnalysis as mda
     from espressomd import MDA_ESP
-    skipIfMissingPythonPackage = ut.case._id
+    skipIfMissingPythonPackage = utx._id
```
```suggestion
    skipIfMissingPythonPackage = lambda x: x
```

KaiSzuttor replied:
this would be even more readable, right?

jngrad replied:
we want to hide implementation details of python decorators (that's why `unittest` created `_id()` in the first place)

also, that lambda expression would be transformed into a regular function by autopep8

KaiSzuttor reviewed (commented):
LGTM, please clarify the private id function which must have been introduced elsewhere

jngrad committed:
tests: Make decorator function public

jngrad committed:
python: Fix broken LB getters

jngrad committed:
tests: Add missing feature guard

jngrad committed:
tests: Add more LB properties checks

christophlohrmann reviewed (approved):
Changes after Kai's review look good to me

kodiakhq[bot] committed:
Merge branch 'python' into lb_interface_coverage
